// gmll/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Instance = exports.nbt = exports.files = exports.wrapper = exports.handler = exports.downloader = exports.config = exports.init = void 0;
const tslib_1 = require("tslib");
/**
 * ---------------------------------------------------------------------------------------------
 * INDEX
 * ---------------------------------------------------------------------------------------------
 */
const config_js_1 = require("./modules/config.js");
/**Does a range of required preflight checks. Will cause errors if ignored!*/
async function init() { await (0, config_js_1.initialize)(); }
exports.init = init;
/**The core config class. Used to change the locations of files and to get the location of files as well! */
exports.config = tslib_1.__importStar(require("./modules/config.js"));
/**The main download manager in GMLL. */
exports.downloader = tslib_1.__importStar(require("./modules/downloader.js"));
/**Stuff related to the version and manifest files. Used to install forge, get a complete list of manifest files and so much more! */
exports.handler = tslib_1.__importStar(require("./modules/handler.js"));
/**Integration with other libs */
exports.wrapper = tslib_1.__importStar(require("./modules/wrapper.js"));
/**Provides access to GMLL's file handler */
exports.files = tslib_1.__importStar(require("./modules/objects/files.js"));
/**Provides access to the nbt data reader in gmll */
exports.nbt = tslib_1.__importStar(require("./modules/nbt.js"));
/**The main instance class in gmll */
const instance_js_1 = tslib_1.__importDefault(require("./modules/objects/instance.js"));
exports.Instance = instance_js_1.default;
