// minecraft-launcher-lib/lib/components/m-version/m-version-downloads/m-version-downloads.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MVersionDownloads = void 0;

var _artifact = require("../../artifact");

class MVersionDownloads {
  static from(child, parent) {
    if (!parent) {
      if (child instanceof MVersionDownloads) return child;
      parent = {};
    }

    const {
      client: parentClient = {},
      server: parentServer = {}
    } = parent;
    const {
      client = parentClient,
      server = parentServer
    } = child;
    return new MVersionDownloads(_artifact.Artifact.from(client, {
      path: 'client.jar'
    }), _artifact.Artifact.from(server, {
      path: 'server.jar'
    }));
  }

  constructor(client, server) {
    this._client = _artifact.Artifact.from(client);
    this._server = _artifact.Artifact.from(server);
  }

  get client() {
    return this._client;
  }

  set client(client) {
    this._client = _artifact.Artifact.from(client);
  }

  get server() {
    return this._server;
  }

  set server(server) {
    this._server = _artifact.Artifact.from(server);
  }

  toJSON() {
    const {
      client,
      server
    } = this;
    return {
      client,
      server
    };
  }

}

exports.MVersionDownloads = MVersionDownloads;
